Source/Layout/ASYogaLayoutSpec.mm

```diff
@@ -0,0 +1,162 @@
+//
+//  ASYogaLayoutSpec.mm
+//  AsyncDisplayKit
+//
+//  Created by Scott Goodson on 5/6/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <AsyncDisplayKit/ASAvailability.h>
+
+#if YOGA /* YOGA */
+#if !YOGA_TREE_CONTIGUOUS /* !YOGA_TREE_CONTIGUOUS */
+
+#import <AsyncDisplayKit/ASYogaLayoutSpec.h>
+#import <AsyncDisplayKit/ASYogaUtilities.h>
+#import <AsyncDisplayKit/ASDisplayNode+Beta.h>
+#import <AsyncDisplayKit/ASLayout.h>
+#import <AsyncDisplayKit/ASLayoutSpec+Subclasses.h>
+
+#define YOGA_LAYOUT_LOGGING 0
+
+@implementation ASYogaLayoutSpec
+
+- (ASLayout *)layoutForYogaNode:(YGNodeRef)yogaNode
+{
+  BOOL isRootNode = (YGNodeGetParent(yogaNode) == NULL);
+  uint32_t childCount = YGNodeGetChildCount(yogaNode);
+
+  NSMutableArray *sublayouts = [NSMutableArray arrayWithCapacity:childCount];
+  for (uint32_t i = 0; i < childCount; i++) {
+    [sublayouts addObject:[self layoutForYogaNode:YGNodeGetChild(yogaNode, i)]];
+  }
+
+  id <ASLayoutElement> layoutElement = (__bridge id <ASLayoutElement>)YGNodeGetContext(yogaNode);
+  CGSize size = CGSizeMake(YGNodeLayoutGetWidth(yogaNode), YGNodeLayoutGetHeight(yogaNode));
+
+  if (isRootNode) {
+    // The layout for root should have position CGPointNull, but include the calculated size.
+    return [ASLayout layoutWithLayoutElement:layoutElement size:size sublayouts:sublayouts];
+  } else {
+    CGPoint position = CGPointMake(YGNodeLayoutGetLeft(yogaNode), YGNodeLayoutGetTop(yogaNode));
+    // TODO: If it were possible to set .flattened = YES, it would be valid to do so here.
+    return [ASLayout layoutWithLayoutElement:layoutElement size:size position:position sublayouts:nil];
+  }
+}
+
+- (void)setupYogaNode:(YGNodeRef)yogaNode forElement:(id <ASLayoutElement>)element withParentYogaNode:(YGNodeRef)parentYogaNode
+{
+  ASLayoutElementStyle *style = element.style;
+
+  YGNodeSetContext(yogaNode, (__bridge void *)element);
```

Comment:
Small issue, not worth blocking merge. Thinking about ownership of `element`.

Should this be something like `(__bridge_retained void *)element` then in the yoga context destructor (YGFree?), `id val = (__bridge_transfer id)YGNodeGetContext(node)`?

Again, this looks like a headache that we don't need to wrestle with unless we see crashes from deallocated elements in a particular version of the framework. But now it's written down 🙂 .